distinguish mutability of Storage::unsafe_data

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98162
* #98163
* #97874
* #97804
* #97776
* #97744
* #97743
* #97694
* #97690
* __->__ #98152

See D44409928.

Differential Revision: [D44426197](https://our.internmc.facebook.com/intern/diff/D44426197/)

Commit:
distinguish mutability of Storage::unsafe_data

See D44409928.

Differential Revision: [D44426197](https://our.internmc.facebook.com/intern/diff/D44426197/)

[ghstack-poisoned]

Commit:
Update on "distinguish mutability of Storage::unsafe_data"


See D44409928.

Differential Revision: [D44426197](https://our.internmc.facebook.com/intern/diff/D44426197/)

[ghstack-poisoned]

Commit:
Update on "distinguish mutability of Storage::unsafe_data"


See D44409928.

Differential Revision: [D44426197](https://our.internmc.facebook.com/intern/diff/D44426197/)

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1555,7 +1556,8 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   template <typename T>
   inline T* unsafe_data() const {
-    return storage_.unsafe_data<T>() + storage_offset_;
+    return storage_.unsafeGetStorageImpl()->mutable_unsafe_data<T>() +
```
Are we going to put the mutable variant on Storage? Seems advisable ...

ezyang commented on torch/csrc/Storage.cpp:
```diff
@@ -164,14 +164,15 @@ static PyObject* THPStorage_pynew(
             /*resizable=*/true));
     THPObjectPtr item;
+
+    c10::Storage storage = *std::move(self->cdata);
```
Why do you need to move the storage out of self temporarily 🤔

ezyang replied:
If this is because of const correctness, this is misguided, for the same reason const Tensor& should be able to vend both mutable and immutable data ptrs

Commit:
Update on "distinguish mutability of Storage::unsafe_data"


See D44409928.

Differential Revision: [D44426197](https://our.internmc.facebook.com/intern/diff/D44426197/)

[ghstack-poisoned]

ezyang commented on torch/csrc/Storage.cpp:
```diff
@@ -234,9 +236,10 @@ static PyObject* THPStorage_get(THPStorage* self, PyObject* index) {
-
     at::StorageImpl* old_storage_impl = storage.unsafeGetStorageImpl();
+
+    uint8_t* data = old_storage_impl->mutable_unsafe_data<uint8_t>();
```
I think this change is ok, but I don't like diffs where something safe goes to unsafe. Needs more explanation

ezyang reviewed (changes requested):
It seems to me that more accessors on Storage should be added first.

I also think the "unsafe" nomenclature on the storage data ptr accessor should go. There is no way to actually check that the dtypes match; let's not force people to say unsafe when there isn't a safe alternative

dagitses commented:
> It seems to me that more accessors on Storage should be added first.
> 
> I also think the "unsafe" nomenclature on the storage data ptr accessor should go. There is no way to actually check that the dtypes match; let's not force people to say unsafe when there isn't a safe alternative

Yeah I agree it's not ideal. But it looks like, at least for the non-const pointers, it's only ever used with uint8 or char as the types. char is definitely safe, uint8_t is not legal, but close enough and could probably be converted to std::byte.

So what I would propose:
 1) drop the unsafe variants
 2) eschew the template in favor of accessors `bytes() -> const std::byte*` and `mutable_bytes() -> std::byte*`.
 3) eventually merge the `data()` and `bytes()` accessors. I don't think we really need both.

And then only have the safe/unsafe variants at the Tensor level, where we actually have enough information to distinguish them.

dagitses commented:
> 

I take it back. Storage and StorageImpl should only have `data() -> const char*` and `mutable_data() -> char*`. There's no need to integrate interpreting them as any type into the API.

dagitses commented:
other changes are done and merged. abandoning this.
